### PR TITLE
[dask] Add type hints.

### PR DESCRIPTION
**Pull request by trivialfis**
Related https://github.com/dmlc/xgboost/issues/6496 .

* Add type hints to dask interface.
* Add a `py.typed` file to python package.  See https://www.python.org/dev/peps/pep-0561/#id18 for details.
* Add various parameters to match the interface with single node.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -174,7 +174,8 @@ jobs:
         python -m pip install wheel setuptools mypy
     - name: Run mypy
       run: |
-        cd python-package && mypy . || true
+        cd python-package
+        mypy ./xgboost/dask.py ../tests/python/test_with_dask.py --follow-imports=silent
```
Comment (hcho3) [inline .github/workflows/main.yml:178]: Can we still run mypy on other Python files? (We can keep it nonblocking)

Comment (trivialfis) [reply .github/workflows/main.yml:178]: Yes.  Let me do that.

**Comment by trivialfis** [comment]
The html doc becomes quite verbose though ...

**Comment by hcho3** [comment]
@trivialfis We can hide the type hints from the doc by using `autodoc_typehints` option in Sphinx: https://www.sphinx-doc.org/en/2.0/usage/extensions/autodoc.html#confval-autodoc_typehints. WDYT?

**Comment by trivialfis** [comment]
Thanks for the reference, let me test it later.

**Comment by trivialfis** [comment]
@hcho3 Do you happen to know how to use smaller font size for API reference?  I compared the doc for xgboost and other projects and the font size in xgboost's doc is significantly larger.  If not then I will continue googling ...

**Comment by hcho3** [comment]
@trivialfis This is due to the choice of the theme. I'm not sure if we should necessarily follow other projects when it comes to the font size. I think our doc is similar to sklearn when it comes to the font size.

**Comment by hcho3** [comment]
@trivialfis Why don't we consider splitting our API doc into multiple pages? Right now, it's a single page with lots of functions and modules, and it might be nice to break it up.

**Comment by trivialfis** [comment]
As can be seen from the screen shot, information like `Optional` and `Union` is omitted in the doc strings, which is less accurate than the checked type hints.  It would be difficult for us to get everything right without proper tools like mypy.

**Comment by hcho3** [comment]
@trivialfis https://github.com/agronholm/sphinx-autodoc-typehints seems promising. It should make the type hints appear in the docstring.

**Comment by trivialfis** [comment]
Let me try it.

**Comment by trivialfis** [comment]
Debugged it a little bit didn't get it to work.  It has some weird issue with import `No module named 'pandas.util'; 'pandas' is not a package`.  The error happens inside dask and happens only when building doc.  Also weirdly if I try to print out the backtrace inside dask, the doc build passed but non of the type hints is used.

**Comment by hcho3** [comment]
@trivialfis I found a way to move the type hints to the description, as follows:
Screen Shot 2020-12-28 at 8 30 53 PM

I found the trick from Optuna (optuna/optuna#2147), which actually moved the type hints to the description area 5 days ago.

**Review by hcho3 — APPROVED** [review]
LGTM, as long as tests pass.

**Pull request merged**